### ISSUE TITLE
No llama3_features_tensor.py provided

### ISSUE THREAD

**Issue by Hugo-Zhu** [BUPT-SN/ETM#1]
No llama3_features_tensor.py file provided

**Comment by TianSongS** [comment]
AI辅助实现的
`import os
os.environ["CUDA_VISIBLE_DEVICES"] = "7"
import torch
import pandas as pd
import numpy as np
import re
import os
from transformers import AutoTokenizer, AutoModel
from tqdm import tqdm

# --- Constants based on train.py and paper ---
LLAMA_MODEL_NAME = "Meta-Llama-3-8B-Instruct"

# Default paths, can be overridden by command-line arguments
DEFAULT_CSV_PATH = "./kaggle/mbti_1.csv"  # As in train.py
DEFAULT_OUTPUT_DIR = "./features_tensor/"  # As in train.py
DEFAULT_OUTPUT_FILENAME = "features_tensor_5.pt" # As in train.py, _5 likely means dm=5

SEED = 29  # From train.py's seed_value
DM = 5  # Number of last layers to average, from paper Table 5 ($d_m=5$)
LLAMA_MAX_TOKEN_LEN = 8192 # Llama 3 context window, tokenizer default

# --- Text Preprocessing (adapted and corrected from train.py) ---
MBTIs_tuple = ('INTJ', 'INTP', 'INFP', 'ENTP', 'ISTP', 'ISFP', 'ESTJ', 'ISTJ', 'ESTP', 'ISFJ', 'ENFP', 'ESFP', 'ESFJ', 'ENFJ', 'INFJ', 'ENTJ')
TOKEN_MASK = '<mask>'  # 'token' in train.py

def find_all_MBTIs_lower(post_lower: str, mbti_list: tuple) -> list:
    """Finds all occurrences of MBTI types in a lowercased post."""
    found_indices = []
    for mbti_type in mbti_list:
        for match in re.finditer(mbti_type.lower(), post_lower):
            found_indices.append((match.start(), match.end()))
    found_indices.sort()
    return found_indices

def text_preprocessing_for_llama(text: str) -> str:
    """
    Processes text for Llama3 embedding generation, mirroring train.py's preprocessing.
    This includes lowercasing, cleaning, and MBTI type masking.
    """
    if not isinstance(text, str):
        text = str(text)

    text_lower = text.lower()

    text_lower = re.sub(r'\s+', ' ', text_lower)
    text_lower = re.sub(r'https?://\S+|www\.\S+', '', text_lower)
    text_lower = re.sub(r'<.*?>', '', text_lower)
    text_lower = re.sub(r'\n', ' ', text_lower)
    text_lower = re.sub(r'\w*\d\w*', '', text_lower)

    mbti_indices = find_all_MBTIs_lower(text_lower, MBTIs_tuple)
    
    if mbti_indices:
        processed_parts = []
        current_pos = 0
        for start, end in mbti_indices:
            if start >= current_pos:
                processed_parts.append(text_lower[current_pos:start])
                processed_parts.append(TOKEN_MASK)
                current_pos = end
        processed_parts.append(text_lower[current_pos:])
        text_lower = "".join(processed_parts)

    text_lower = text_lower.encode('ascii', 'ignore').decode('ascii')
    
    if text_lower.startswith("'") and text_lower.endswith("'"):
        text_lower = text_lower[1:-1]
    
    text_lower = re.sub(r'\s+', ' ', text_lower).strip()
    
    return text_lower

def generate_llama_embeddings(csv_path: str, output_dir: str, output_filename: str):
    """
    Generates Llama3 embeddings for the 'posts' column in the provided CSV,
    shuffles the data identically to train.py, and saves the embeddings.
    """
    print(f"Starting Llama3 embedding generation...")
    print(f"Configuration: SEED={SEED}, DM={DM}")

    np.random.seed(SEED)
    # torch.manual_seed(SEED) # For torch operations if any were stochastic before model

    print(f"Loading Llama3 model: {LLAMA_MODEL_NAME}")
    tokenizer = AutoTokenizer.from_pretrained(LLAMA_MODEL_NAME)
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token 
        
    model = AutoModel.from_pretrained(
        LLAMA_MODEL_NAME,
        output_hidden_states=True,
        torch_dtype=torch.bfloat16,
        device_map="auto"
    )
    model.eval()
    model_device = model.device # Or a specific device if device_map is not "auto"
    print(f"Llama3 model loaded. Primary device: {model_device}")

    print(f"Loading data from: {csv_path}")
    try:
        data_df = pd.read_csv(csv_path)
    except FileNotFoundError:
        print(f"ERROR: CSV file not found at {csv_path}. Please check the path.")
        return

    print(f"Shuffling data with random_state={SEED}...")
    data_df_shuffled = data_df.sample(frac=1, random_state=SEED).reset_index(drop=True)
    print(f"Data loaded and shuffled. Total samples: {len(data_df_shuffled)}")

    all_llama_embeddings = []
    print("Generating embeddings for each sample...")

    for _, row in tqdm(data_df_shuffled.iterrows(), total=len(data_df_shuffled), desc="Processing samples"):
        user_posts_string = row['posts']
        processed_text = text_preprocessing_for_llama(user_posts_string)
        
        inputs = tokenizer(
            processed_text, 
            return_tensors="pt", 
            truncation=True, 
            # Using a slightly reduced max_length to allow for special tokens if tokenizer adds them
            max_length=LLAMA_MAX_TOKEN_LEN - 20 if LLAMA_MAX_TOKEN_LEN > 20 else LLAMA_MAX_TOKEN_LEN,
            padding=False 
        )
        inputs = {k: v.to(model_device) for k, v in inputs.items()}

        with torch.no_grad():
            outputs = model(**inputs)
        
        hidden_states = outputs.hidden_states # Tuple, length num_layers + 1 (e.g., 33 for Llama-3-8B)
        
        # Select hidden states from the last DM layers (e.g., layers 28-32 for DM=5 in a 32-layer model)
        # hidden_states[-1] is the output of the last layer (e.g., layer 32)
        # hidden_states[-DM:] gives the outputs of the last DM layers
        selected_layer_hidden_states = hidden_states[-DM:]
        
        layer_pooled_vectors = []
        for layer_hs in selected_layer_hidden_states:
            # layer_hs shape: (1, sequence_length, hidden_dimension)
            pooled_vec = torch.mean(layer_hs, dim=1) # Result: (1, hidden_dimension)
            layer_pooled_vectors.append(pooled_vec)
            
        if layer_pooled_vectors:
            stacked_layer_vectors = torch.stack(layer_pooled_vectors) # Shape (DM, 1, hidden_dim)
            final_user_embedding = torch.mean(stacked_layer_vectors, dim=0) # Shape (1, hidden_dim)
            all_llama_embeddings.append(final_user_embedding.squeeze().cpu()) # Squeeze to (hidden_dim)
        else:
            # Fallback, should ideally not be reached if DM > 0
            print(f"Warning: No layer vectors for a sample. Appending zeros.")
            # Dimension 4096 is from train.py's llama_embed_dim
            all_llama_embeddings.append(torch.zeros(4096, dtype=torch.bfloat16)) 

    if not all_llama_embeddings:
        print("No embeddings were generated. Exiting.")
        return
        
    stacked_embeddings = torch.stack(all_llama_embeddings)
    print(f"Embeddings tensor shape: {stacked_embeddings.shape}")

    os.makedirs(output_dir, exist_ok=True)
    output_path = os.path.join(output_dir, output_filename)
    
    print(f"Saving embeddings to: {output_path}")
    torch.save(stacked_embeddings, output_path)
    print("Embeddings saved successfully.")

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="Generate Llama3 embeddings for MBTI dataset.")
    parser.add_argument(
        "--csv_path", 
        type=str, 
        default=DEFAULT_CSV_PATH, 
        help=f"Path to the input CSV file (default: {DEFAULT_CSV_PATH})."
    )
    parser.add_argument(
        "--output_dir", 
        type=str, 
        default=DEFAULT_OUTPUT_DIR, 
        help=f"Directory to save the output tensor (default: {DEFAULT_OUTPUT_DIR})."
    )
    parser.add_argument(
        "--output_filename", 
        type=str, 
        default=DEFAULT_OUTPUT_FILENAME, 
        help=f"Filename for the output tensor (default: {DEFAULT_OUTPUT_FILENAME})."
    )
    
    args = parser.parse_args()
    
    generate_llama_embeddings(args.csv_path, args.output_dir, args.output_filename)`
